src/v/cloud_storage/recursive_directory_walker.cc

```diff
@@ -138,17 +139,51 @@ ss::future<walk_result> recursive_directory_walker::walk(
 
     fragmented_vector<ss::sstring> empty_dirs;
 
+    // Listing directories involves blocking I/O which in Seastar is serviced by
+    // a dedicated thread pool and workqueue. When listing directories with
+    // large number of sub-directories this leads to a large queue of work items
+    // for each of which we have to incur the cost of task switching. Even for a
+    // lightly loaded reactor, it can take up to 1ms for a full round-trip. The
+    // round-trip time is ~100x the syscall duration. With 2 level directory
+    // structure, 100K files, 2 getdents64 calls per directory this adds up to
+    // 400K syscalls and 6m of wall time.
+    //
+    // By running the directory listing in parallel we also parallelize the
+    // round-trip overhead. This leads to a significant speedup in the
+    // directory listing phase.
+    //
+    // Limit the number of concurrent directory reads to avoid running out of
+    // file descriptors.
+    //
+    // Empirical testing shows that this value is a good balance between
+    // performance and resource usage.
+    const size_t max_concurrency = 1000;
+
+    std::vector<ss::sstring> targets;
```

Comment:
nit: Should we use a `fragmented_vector` here?

Reply:
That's 16KiB at most using current config constraints. I tend to default to std data structures unless I know we'll run into problems. Don't have much data to back this preference though.

Reply:
yeh, it seems ok especialy since the max size is controlled by that constant above